### PR TITLE
Feature/introduce stepper motor api utilizing a "motor" bus

### PR DESCRIPTION
**Pull request by jilaypandya**
https://github.com/zephyrproject-rtos/zephyr/issues/68547

```
    &spi_controller { // bus:spi
      status = "okay";

      my_motor_controller: my_motor_controller@0 { // bus:motor & on-bus:spi
        status = "okay";
        compatible = "my_motor,controller";
        reg = <0>;
        #address-cells = <1>;
        #size-cells = <0>;

        my_motor: my_motor_0@0{ //on-bus: motor
          status = "okay";
          compatible = "tmc5041-stepper-motor-device";
          reg=<0>;
          gear-ratio = "1.234567";
          micro-step-res = <256>;
          steps-per-revolution = <200>;
          stall-guard-setting = <6>;
          controller-spec-reg-settings = <SPEC_REG_ADD_FOR_MOTOR_0(0)  SPEC_REG_VALUE_FOR_MOTOR_0>;
        };

        my_motor: my_motor_1@1{ //on-bus: motor
          status = "okay";
          compatible = "tmc5041-stepper-motor-device";
          reg=<1>;
          gear-ratio = "9.876543";
          micro-step-res = <128>;
          steps-per-revolution = <200>;
          stall-guard-setting = <0>;
          controller-spec-reg-settings = <SPEC_REG_ADD_FOR_MOTOR_1(1)  SPEC_REG_VALUE_FOR_MOTOR_1>;
        };
      };
    };
```

### CODE REVIEW

File: dts/bindings/stepper_motor/stepper-motor-device.yaml [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 Carl Zeiss Meditec AG
+# SPDX-License-Identifier: Apache-2.0
+
+description: Stepper Motor Device
+
+include: base.yaml
+
+on-bus: motor
```
Comment (gmarull) [inline dts/bindings/stepper_motor/stepper-motor-device.yaml:8]: this looks wrong to me. Where is this bus standardized?

File: dts/bindings/stepper_motor/stepper-motor-device.yaml [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+compatible: "stepper-motor-device"
+
+properties:
+
+  friendly-name:
+    type: string
+    description: |
+      Human readable string describing the stepper motor. It can be used to
+      distinguish multiple instances of the same model.
```
Comment (gmarull) [inline dts/bindings/stepper_motor/stepper-motor-device.yaml:18]: a backdoor into `label`?

File: dts/bindings/stepper_motor/stepper-motor-device.yaml [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+  gear-ratio:
+    type: string
+    default: "1.0"
+    description: |
+      Gear ratio from stepper motor to end transmission, This value can be in float and hence
+      the property type string
```
Comment (gmarull) [inline dts/bindings/stepper_motor/stepper-motor-device.yaml:25]: please avoid pseudo floating points in DT, just specify the ratio with numerator/denominator, see eg voltage divider

File: dts/bindings/stepper_motor/stepper-motor-device.yaml [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      - 2
+      - 4
+      - 8
+      - 16
+      - 32
+      - 64
+      - 128
+      - 256
+    description: |
+      Enter the microstep resolution from one of the enum values
```
Comment (gmarull) [inline dts/bindings/stepper_motor/stepper-motor-device.yaml:40]: isn't this specialized to tmc5041?

File: dts/bindings/stepper_motor/stepper-motor-device.yaml [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+  steps-per-revolution:
+    type: int
+    default: 200
```
Comment (gmarull) [inline dts/bindings/stepper_motor/stepper-motor-device.yaml:44]: defaults need justification

File: dts/bindings/stepper_motor/stepper-motor-device.yaml [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  controller-spec-reg-settings:
+    type: array
+    description: |
+      This property allows for setting controller-spec-register values directly from the dts.
+      The driver will iterate through this array and will treat all even elements are reg_addresses
+      and odd elements as reg_values.
+
+      Note: Enter key-value pairs as follows
+
+      controller-spec-reg-settings = <reg_a_address reg_a_value>, <reg_b_address, reg_b_value>;
```
Comment (gmarull) [inline dts/bindings/stepper_motor/stepper-motor-device.yaml:58]: this looks wrong, specially in a file that is supposed to be generic.

File: dts/bindings/stepper_motor/tmc5041-stepper-motor-device.yaml [hunk 7]
```diff
@@ -0,0 +1,21 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 Carl Zeiss Meditec AG
+# SPDX-License-Identifier: Apache-2.0
+
+description: TMC5041 based Stepper Motor Device
+
+include: stepper-motor-device.yaml
+
+compatible: "tmc5041-stepper-motor-device"
```
Comment (gmarull) [inline dts/bindings/stepper_motor/tmc5041-stepper-motor-device.yaml:8]: wrong compatible format, please, check DT documentation

File: include/zephyr/drivers/stepper_motor_controller.h [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+typedef int32_t (*stepper_motor_controller_reg_read_t)(const struct device *dev,
+						       uint8_t reg_address, uint32_t *reg_value);
+__subsystem struct stepper_motor_controller_api {
+	stepper_motor_controller_reset_t stepper_motor_controller_reset;
+	stepper_motor_controller_write_t stepper_motor_controller_write;
+	stepper_motor_controller_read_t stepper_motor_controller_read;
+
+	stepper_motor_controller_reg_write_t stepper_motor_controller_write_reg;
+	stepper_motor_controller_reg_read_t stepper_motor_controller_read_reg;
+};
```
Comment (gmarull) [inline include/zephyr/drivers/stepper_motor_controller.h:118]: this really looks like a kind of MFD device, where you have a generic read/write for multiple elements of the same chip. I think this modeling of a bus is really an overkill. The API could just be multichannel, or, just have a parent MFD and as child nodes as motors the chip has.

File: include/zephyr/drivers/stepper_motor_device.h [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+ * @param calibrate_func Pointer to the calibration function
+ *
+ * @return 0 if successful, negative code if failure
+ */
+static inline int32_t stepper_motor_register_calibrate_func(
+	const struct device *dev,
+	int32_t (*calibrate_func)(const struct device *dev, int32_t *min_pos, int32_t *max_pos))
+{
+	return z_impl_stepper_motor_register_calibrate_func(dev, calibrate_func);
+}
```
Comment (gmarull) [inline include/zephyr/drivers/stepper_motor_device.h:358]: not sure I follow here

**Review by gmarull — CHANGES_REQUESTED** [review]
I think that the bus modelling is just wrong, can be done with a multi-channel API, or, as a kind of MFD device, or even just internal generic read/write calls inside the same multi-instance driver (ie, nothing needed).

Btw, there's an ancient PR from me adding an API for stepper motors which also included async API and an implementation using STM32 peripherals. It's likely outdated and not feature complete, but it may be useful to see if the API could be used for that implementation. https://github.com/zephyrproject-rtos/zephyr/pull/26221

**Comment by jilaypandya** [comment]
Hi @gmarull Thanks for the review. 

I like your approach as well and have created a seperate PR https://github.com/zephyrproject-rtos/zephyr/pull/68774 following the same. However, one of the downsides with the approach is that the motors dont get listed as devices. 

Hence for example with the following device tree 
```
&spi-controller {
    status = "okay";
    tmc5041: tmc5041@0 {
        status = "okay";
        compatible = "trinamic,tmc5041";
        reg = <0>;
        #address-cells = <1>;
        #size-cells = <0>;

        motor_0: motor_0@0 {
            status = "okay";
            reg = <0>;
            gear-ratio = "1.234567";
            micro-step-res = <256>;
            steps-per-revolution = <200>;
            stall-guard-setting = <12>;
        };

        motor_1: motor_1@1 {
            status = "okay";
            reg = <1>;
            stall-guard-setting = <34>;
            steps-per-revolution = <200>;
        };
    };
};
```

with bus approach,`stepper_motor_reset(motor_0)` would suffice wheras with the child-binding approach the reset call would be `stepper_motor_reset(tmc5041,0)` i.e. one would always need to know the proper index of the motor.

However, i am open for suggestions and trying out different approaches.

File: dts/bindings/stepper_motor/stepper-motor-device.yaml (hunk 3 above)
Comment (jilaypandya) [reply dts/bindings/stepper_motor/stepper-motor-device.yaml:25]: But why? 

With [DT_STRING_UNQUOTED()](https://docs.zephyrproject.org/apidoc/latest/group__devicetree-generic-prop.html#gad71ae303ce20e116a75c23ca552c2225) you can get a string property's value as an unquoted sequence of tokens.

It removes "the quotes" from string-valued properties. That can be useful, for example, when defining floating point values as a string in devicetree that you would like to use to initialize a float or double variable in C.

File: dts/bindings/stepper_motor/stepper-motor-device.yaml (hunk 4 above)
Comment (jilaypandya) [reply dts/bindings/stepper_motor/stepper-motor-device.yaml:40]: Hmm not really, as far as i know microsteps are always in the power of 2. Here an excerpt from DRV8825
MicrosoftTeams-image

**Comment by jilaypandya** [comment]
Closing this PR in favour of https://github.com/zephyrproject-rtos/zephyr/pull/68774

**Pull request closed without merging**